control/controlbase/conn.go

```diff
@@ -96,8 +97,26 @@ func (c *Conn) readNLocked(total int) ([]byte, error) {
 		if total <= c.rx.n {
 			return c.rx.buf[:total], nil
 		}
-
-		n, err := c.conn.Read(c.rx.buf[c.rx.n:])
+		var n int
+		var err error
+		if c.rx.buf == nil && c.rx.n == 0 && total == headerLen {
```

Comment:
This branch is suspect: if one of the latter two conditions is false, the else would end up reading into a nil buffer. So either the latter conditions aren't necessary and are always true here, or there are more alternatives needed?

Reply:
Tightened it up. PTAL.